sockets/unix_socket.go

```diff
@@ -79,7 +79,7 @@ func WithChmod(mask os.FileMode) SockOption {
 
 // NewUnixSocketWithOpts creates a unix socket with the specified options
 func NewUnixSocketWithOpts(path string, opts ...SockOption) (net.Listener, error) {
-	if err := syscall.Unlink(path); err != nil && !os.IsNotExist(err) {
+	if err := os.Remove(path); err != nil && !os.IsNotExist(err) {
```

Comment:
Thanks, so, this patch would not help getting into the situation (which is due to a race-condition between containers starting and the API being up), but would help _recovering_ from that situation (by being able to remove the faulty directory instead of producing an error. Obviously, if there's containers still running that mount `/var/run/docker.sock` that would likely be problematic still 🤔 

Change itself seems fine to me; if `path` is a directory it would attempt to remove it, which should be fairly safe, as it would fail if the directory is not empty, so there should be no real risk of accidentally removing files that we don't want to.

Reply:
@thaJeztah @tiborvass thanks.

For creating the dir of /var/run/docker.sock, the root cause of this is that dockerd can still handle the API when it shutdown, at that time,  dockerd create a container and share the host path /var/run/docker.sock, but it is not exist，so dockerd create it. I have a pr https://github.com/moby/moby/pull/38241 abort this，PTAL again, thanks very much.